core/ipc/rpc/kafka/src/main/java/org/opennms/core/ipc/rpc/kafka/KafkaRpcClientFactory.java

```diff
@@ -0,0 +1,311 @@
+                long timeout) {
+            this.responseFuture = responseFuture;
+            this.rpcModule = rpcModule;
+            this.expirationTime = timeout;
+            this.rpcId = rpcId;
+        }
+
+        @Override
+        public void sendResponse(String message) {
+            T response = null;
+            // When message is not null, it's called from kafka consumer otherwise it is from timeout tracker.
+            if (message != null) {
+                response = rpcModule.unmarshalResponse(message);
+                if (response.getErrorMessage() != null) {
+                    responseFuture.completeExceptionally(new RemoteExecutionException(response.getErrorMessage()));
+                } else {
+                    responseFuture.complete(response);
+                }
+            } else {
+                responseFuture.completeExceptionally(new RequestTimedOutException(new TimeoutException()));
+            }
+            rpcResponseMap.remove(rpcId);
+        }
+
+        @Override
+        public int compareTo(Delayed other) {
+            long myDelay = getDelay(TimeUnit.MILLISECONDS);
+            long otherDelay = other.getDelay(TimeUnit.MILLISECONDS);
+            if (myDelay < otherDelay)
+                return -1;
+            if (myDelay == otherDelay)
+                return 0;
+            return 1;
+        }
+
+        @Override
+        public long getDelay(TimeUnit unit) {
+            long now = System.currentTimeMillis();
+            return unit.convert(expirationTime - now, TimeUnit.MILLISECONDS);
+        }
+
+    }
+
+    private class KafkaConsumerRunner implements Runnable {
+        private final KafkaConsumer<String, byte[]> consumer;
+        private final AtomicBoolean closed = new AtomicBoolean(false);
+
+        public KafkaConsumerRunner(KafkaConsumer<String, byte[]> consumer) {
+            this.consumer = consumer;
+        }
+
+        @Override
+        public void run() {
+            try {
+                // subscribe to topics with "rpc-response" as part of topic name.
+                Pattern pattern = Pattern.compile(".*rpc-response.*");
+                consumer.subscribe(pattern);
+                while (!closed.get()) {
+                    ConsumerRecords<String, byte[]> records = consumer.poll(Long.MAX_VALUE);
+                    for (ConsumerRecord<String, byte[]> record : records) {
+                        RpcMessageProtos.RpcMessage rpcMessage = RpcMessageProtos.RpcMessage.parseFrom(record.value());
+                        // Get Response callback from rpcId and send rpc content to callback.
+                        ResponseCallback responseCb = rpcResponseMap.get(rpcMessage.getRpcId());
+                        if (responseCb != null) {
+                            responseCb.sendResponse(rpcMessage.getRpcContent().toStringUtf8());
+                        }
```

Comment:
We should log something when we receive a response for an entry that can not be found in the map.